crates/sui-types/src/messages.rs

```diff
@@ -1328,27 +1346,20 @@ impl CertifiedTransaction {
             return Ok(());
         }
 
-        let mut obligation = VerificationObligation::default();
-        self.add_to_verification_obligation(committee, &mut obligation)?;
-        obligation.verify_all().map(|_| ())
-    }
+        let mut obligation: VerificationObligation<AggregateAuthoritySignature>
+            = VerificationObligation::default();
 
-    fn add_to_verification_obligation(
-        &self,
-        committee: &Committee,
-        obligation: &mut VerificationObligation<AggregateAuthoritySignature>,
-    ) -> SuiResult<()> {
-        // Add the obligation of the sender signature verification.
-
-        // // Add the obligation of the authority signature verifications.
         let mut message = Vec::new();
         self.data.write(&mut message);
         let idx = obligation.add_message(message);
-
-        // let idx = self.add_sender_sig_to_verification_obligation(obligation)?;
-
+        
+        if self.add_sender_sig_to_verification_obligation(&mut obligation, idx).is_err() {
+            self.verify_sender_signature()?;
```

Comment:
Try to add to verification obligation. If mismatched type - verify independently.

Reply:
We can pattern-match into the variants of an `enum` and essentially have an accumulating `VerificationObligation` with two (or more) inner accumulators. Then, upon verification, we can discharge each inner accumulator independently.

That's of course if all schemes in those inner accumulators support batch verification. For those which don't, the best we can do is to indeed verify independently, but ideally asynchronously.